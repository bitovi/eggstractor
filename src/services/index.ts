export * from './collection.service';
export * from './token.service';
export * from './variable.service'; 