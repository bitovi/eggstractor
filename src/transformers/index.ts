export * from "./scss.transformer";
export * from "./css.transformer";
