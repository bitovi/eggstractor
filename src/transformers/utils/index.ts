export * from './deduplicate-messages.utils';
export * from './get-style-property-and-value.utils';
export * from './group-by.utils';
export * from './update-padding-based-on-border.utils';
