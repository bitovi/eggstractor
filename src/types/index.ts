export * from './tokens';
export * from './processors';
export * from './gradients'; 