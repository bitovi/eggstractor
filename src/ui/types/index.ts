/**
 * UI-specific types for the Eggstractor plugin
 * Extends base types from @types for UI components
 */

export interface PluginConfig {
  githubToken: string;
  repoPath: string;
  filePath: string;
  branchName: string;
  format: 'scss' | 'css' | 'tailwind-scss' | 'tailwind-v4';
}

export interface PluginState {
  isGenerating: boolean;
  generatedContent: string | null;
  warnings: string[];
  errors: string[];
  progress: ProgressState | null;
}

export interface ProgressState {
  percentage: number;
  message: string;
}

export interface PluginMessage {
  type: string;
  [key: string]: any;
}

export interface PluginInterfaceProps extends PluginState {
  config: PluginConfig;
  onConfigChange: (config: Partial<PluginConfig>) => void;
  onGenerateStyles: () => void;
  onCreatePR: () => void;
  onExportTestData: () => void;
}

export interface FormFieldProps {
  id: string;
  label: string;
  type: 'text' | 'password' | 'select';
  value: string;
  placeholder?: string;
  options?: Array<{ value: string; label: string }>;
  onChange: (value: string) => void;
}

export interface ButtonProps {
  children: React.ReactNode;
  onClick: () => void;
  disabled?: boolean;
  variant?: 'primary' | 'secondary';
  loading?: boolean;
}

export interface ProgressBarProps {
  progress: ProgressState;
}

export interface StatusDisplayProps {
  warnings: string[];
  errors: string[];
  generatedContent: string | null;
}
