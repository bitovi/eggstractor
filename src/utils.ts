export default {
  toBase64(str: string): string {
    const base64chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789+/';
    const utf8str = encodeURIComponent(str).replace(/%([0-9A-F]{2})/g,
      (_, p1) => String.fromCharCode(parseInt(p1, 16))
    );
    let i = 0;
    let result = '';
    while (i < utf8str.length) {
      const char1 = utf8str.charCodeAt(i++);
      const char2 = i < utf8str.length ? utf8str.charCodeAt(i++) : NaN;
      const char3 = i < utf8str.length ? utf8str.charCodeAt(i++) : NaN;
      const enc1 = char1 >> 2;
      const enc2 = ((char1 & 3) << 4) | (char2 >> 4);
      const enc3 = ((char2 & 15) << 2) | (char3 >> 6);
      const enc4 = char3 & 63;
      result += base64chars[enc1] + base64chars[enc2] +
        (isNaN(char2) ? '=' : base64chars[enc3]) +
        (isNaN(char3) ? '=' : base64chars[enc4]);
    }
    return result;
  },
  sanitizeName(name: string): string {
    return name.toLowerCase().replace(/[^a-z0-9]/g, '-');
  },
  sanitizeSegment(segment: string): string {
    return segment.toLowerCase().replace(/[^a-z0-9]/g, '-');
  },
  rgbToHex(r: number, g: number, b: number): string {
    const rHex = Math.round(r * 255).toString(16).padStart(2, '0');
    const gHex = Math.round(g * 255).toString(16).padStart(2, '0');
    const bHex = Math.round(b * 255).toString(16).padStart(2, '0');
    return '#' + rHex + gHex + bHex;
  },
  rgbaToString(r: number, g: number, b: number, a: number): string {
    const rHex = Math.round(r * 255).toString(16).padStart(2, '0');
    const gHex = Math.round(g * 255).toString(16).padStart(2, '0');
    const bHex = Math.round(b * 255).toString(16).padStart(2, '0');
    const aHex = Math.round(a * 255).toString(16).padStart(2, '0');
    
    return a === 1 ? 
      '#' + rHex + gHex + bHex : 
      '#' + rHex + gHex + bHex + aHex;
  },
  groupBy<T>(arr: T[], key: (item: T) => string): Record<string, T[]> {
    return arr.reduce((groups, item) => {
      const k = key(item);
      if (!groups[k]) groups[k] = [];
      groups[k].push(item);
      return groups;
    }, {} as Record<string, T[]>);
  }
};