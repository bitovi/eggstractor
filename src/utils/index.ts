export * from './color.utils';
export * from './create-naming-convention'
export * from './constants.utils';
export * from './delay.utils';
export * from './get-valid-stylesheet-format.utils';
export * from './node.utils';
export * from './string.utils';
export * from './test.utils';
export * from './units.utils';
export * from './value.utils';
