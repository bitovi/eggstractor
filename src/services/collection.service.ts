import { ComponentSetToken, ComponentToken, TokenCollection } from '../types';
import { getProcessorsForNode } from '../processors';
import { extractComponentSetToken, extractComponentToken, extractNodeToken } from '../services';
import { getNodePathNames } from '../utils/node.utils';

function getFlattenedValidNodes(node: BaseNode): BaseNode[] {
  const result: BaseNode[] = [];

  function traverse(currentNode: BaseNode) {
    // Skip VECTOR and INSTANCE nodes entirely. INSTANCE nodes which are
    // instances of components, which are not relevant for token extraction.
    if ('type' in currentNode && ['VECTOR', 'INSTANCE'].includes(currentNode.type)) {
      return;
    }

    result.push(currentNode);

    if ('children' in currentNode) {
      for (const child of currentNode.children) {
        traverse(child);
      }
    }
  }

  traverse(node);
  return result;
}

export async function collectTokens(onProgress: (progress: number, message: string) => void) {
  const collection: TokenCollection = { tokens: [], components: {}, componentSets: {} };

  let componentToken: ComponentToken | null = null;
  let componentSetToken: ComponentSetToken | null = null;

  let totalNodes = 0;
  let processedNodes = 0;

  let lastTimestamp = Date.now();
  let lastPercentage = -1;

  async function processNode(node: BaseNode) {
    processedNodes++;

    // calculate integer percent in your 10–95 range
    const currentPercentage = Math.floor((processedNodes / totalNodes) * 85) + 10;
    const shouldUpdate = currentPercentage !== lastPercentage || processedNodes === totalNodes;

    if (shouldUpdate) {
      lastTimestamp = currentPercentage;
      onProgress(currentPercentage, `Processing nodes… ${processedNodes}/${totalNodes}`);

      // throttle yields to at most once every 200 ms
      const now = Date.now();
      if (now - lastTimestamp >= 200) {
        await new Promise((r) => setTimeout(r, 0));
        lastTimestamp = now;
      }
    }

    if ('type' in node && 'boundVariables' in node) {
      if (node.type === 'COMPONENT_SET') {
        componentSetToken = extractComponentSetToken(node);
        collection.componentSets[node.id] = componentSetToken;
      }

      if (node.type === 'COMPONENT') {
        componentToken = extractComponentToken(node, componentSetToken!);
        collection.components[node.id] = componentToken;
      }

      const nodePathNames = getNodePathNames(node);
      const processors = getProcessorsForNode(node);

      for (const processor of processors) {
        const tokens = await extractNodeToken(
          node,
          processor,
          nodePathNames,
          componentToken,
          componentSetToken,
        );
        collection.tokens.push(...tokens);
      }
    }
  }

  onProgress(0, 'Loading pages...');
  await figma.loadAllPagesAsync();

  onProgress(5, 'Counting nodes...');
  const validNodes = figma.root.children.flatMap((page) => getFlattenedValidNodes(page));
  totalNodes = validNodes.length;

  onProgress(10, `Processing ${totalNodes} nodes...`);

  // Process all pages in parallel for maximum speed
  const pagePromises = figma.root.children.map(async (page) => {
    const validNodes = getFlattenedValidNodes(page);

    for (const node of validNodes) {
      await processNode(node);
    }
  });

  await Promise.all(pagePromises);

  return collection as Readonly<TokenCollection>;
}
