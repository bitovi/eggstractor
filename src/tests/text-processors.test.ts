import { collectTokens } from '../services';
import { ProcessedValue } from '../types';
import { transformToScss } from '../transformers';
import { fontProcessors } from '../processors/font.processor';
import { createTestVariableResolver } from '../utils/test.utils';
import testData from './fixtures/figma-test-data_paragraph.json';
import testDataAlignment from './fixtures/figma-test-data-alignment.json';
import testDataFontStyle from './fixtures/figma-test-data_font-style.json';

describe('Text Processors', () => {
  describe('Text Align Processor', () => {
    const processor = fontProcessors.filter(p => p.property === 'text-align')[0];

    it('returns null for non-text nodes', async () => {
      const node = {
        type: 'RECTANGLE'
      } as SceneNode;

      const result = await processor.process([], node);
      expect(result).toBeNull();
    });

    it('processes text alignment correctly', async () => {
      const testCases = [
        {
          name: 'center alignment',
          input: {
            type: 'TEXT',
            textAlignHorizontal: 'CENTER',
          } as SceneNode,
          expected: {
            value: 'center',
            rawValue: 'center'
          }
        },
        {
          name: 'right alignment',
          input: {
            type: 'TEXT',
            textAlignHorizontal: 'RIGHT',
          } as SceneNode,
          expected: {
            value: 'right',
            rawValue: 'right'
          }
        },
        {
          name: 'justified alignment',
          input: {
            type: 'TEXT',
            textAlignHorizontal: 'JUSTIFIED',
          } as SceneNode,
          expected: {
            value: 'justify',
            rawValue: 'justify'
          }
        }
      ];

      for (const testCase of testCases) {
        const result = await processor.process([], testCase.input);
        expect(result).toEqual(testCase.expected as ProcessedValue);
      }
    });

    it('should process text alignment correctly', async () => {    
      const pageNode = {
        ...testData,
        type: 'PAGE',
        name: 'text-alignment',
        parent: null,
        width: 100,
        height: 100
      };
  
      const children = testData.children.map((child: BaseNode) => ({
        ...child,
        parent: pageNode,
        width: 100,
        height: 100
      }));
      pageNode.children = children;
  
      // Create variable resolver with complete test data
      const getVariableByIdAsync = await createTestVariableResolver(testData);
  
      // Mock Figma API
      global.figma = {
        currentPage: pageNode,
        variables: {
          getVariableByIdAsync
        }
      };
  
      const tokens = await collectTokens();    
      const { result } = transformToScss(tokens);
  
      expect(result).toMatchSnapshot('text-alignment');
    });
    
    it('should process paragraph alignment correctly', async () => {    
      const pageNode = {
        ...testDataAlignment,
        type: 'PAGE',
        name: 'paragraph-alignment',
        parent: null,
        width: 100,
        height: 100
      };
  
      const children = testDataAlignment.children.map((child: BaseNode) => ({
        ...child,
        parent: pageNode,
        width: 100,
        height: 100
      }));
      pageNode.children = children;
  
      // Create variable resolver with complete test data
      const getVariableByIdAsync = await createTestVariableResolver(testDataAlignment);
  
      // Mock Figma API
      global.figma = {
        currentPage: pageNode,
        variables: {
          getVariableByIdAsync
        }
      };
  
      const tokens = await collectTokens();    
      const { result } = transformToScss(tokens);
  
      expect(result).toMatchSnapshot('paragraph-alignment');
    });

    it('should process font style correctly', async () => {    
      const pageNode = {
        ...testDataFontStyle,
        type: 'PAGE',
        name: 'font-style',
        parent: null,
        width: 100,
        height: 100
      };

      const children = testDataFontStyle.children.map((child: BaseNode) => ({
        ...child,
        parent: pageNode,
        width: 100,
        height: 100
      }));
      pageNode.children = children;
  
      // Create variable resolver with complete test data
      const getVariableByIdAsync = await createTestVariableResolver(testDataFontStyle);
  
      // Mock Figma API
      global.figma = {
        currentPage: pageNode,
        variables: {
          getVariableByIdAsync
        }
      };
  
      const tokens = await collectTokens();    
      const { result } = transformToScss(tokens);
  
      expect(result).toMatchSnapshot('font-style');
    });
  });
}); 