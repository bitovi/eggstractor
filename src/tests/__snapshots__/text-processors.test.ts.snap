// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Text Processors Text Align Processor should process text alignment correctly: text-alignment 1`] = `
"// Generated SCSS Variables
$text-primary: #bfc5c6
$size-16: 16px

// Generated SCSS Mixins
@mixin paragraph-justification-justify
 color: $text-primary
 font-family: Inter
 font-size: 16px
 font-weight: 400
 line-height: normal

@mixin paragraph-justification-right
 color: $text-primary
 font-family: Inter
 font-size: 16px
 font-weight: 400
 line-height: normal

@mixin paragraph-justification-center
 color: $text-primary
 font-family: Inter
 font-size: $size-16
 font-weight: 400
 line-height: normal

@mixin paragraph-justification-left
 color: $text-primary
 font-family: Inter
 font-size: $size-16
 font-weight: 400
 line-height: normal

"
`;
